File: C/SilentVM/Silent_Compiler/Lexer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SilentCompiler
{
    public class Lexer
    {
        List<Tokens> tokens;

        List<string> temp;
        List<string> temp1;

        public List<Tokens> Tokenize(string source, out List<string> values)
        {
            tokens = new List<Tokens>();
            values = new List<string>();
            temp1 = new List<string>();
            temp = source.Replace('\n', ' ').
                Replace('\t', ' ').
                Replace('\r', ' ').
                Replace("+"," + ").
                Replace("-", " - ").
                Replace("/", " / ").
                Replace("*", " * ").
                Replace("==", " == ").
                Replace("!=", " != ").
                Replace(">=", " >= ").
                Replace("<=", " <= ").
                Replace(";", " ; ").
                Replace("(", " ( ").
                Replace(")", " ) ").
                Replace("{", " { ").
                Replace("}", " } ").
                Replace("[", " [ ").
                Replace("]", " ] ").
                Replace(".", " . ").
                Replace(",", " , ").
                Replace("=", " = ").
                Split(' ').ToList();


            for (int i = 0; i < temp.Count; i++)
            {
                if(temp[i] != string.Empty)
                temp1.Add(temp[i]);
            }

            for(int i = 0; i < temp1.Count; i++)
            {
                if (temp1[i] == "public") { tokens.Add(Tokens.Public); }
                else if (temp1[i] == "private") { tokens.Add(Tokens.Private); }
                else if (temp1[i] == "static") { tokens.Add(Tokens.Static); }

                else if (temp1[i] == "namespace") { tokens.Add(Tokens.Namespace); }
                else if (temp1[i] == "using") { tokens.Add(Tokens.Using); }
                else if (temp1[i] == "class") { tokens.Add(Tokens.Class); }
                else if (temp1[i] == ".") { tokens.Add(Tokens.Fullstop); }

                else if (temp1[i] == "method") { tokens.Add(Tokens.Function); }
                else if (temp1[i] == "func") { tokens.Add(Tokens.Function); }

                else if (temp1[i] == "return") { tokens.Add(Tokens.Return); }
                else if (temp1[i] == "struct") { tokens.Add(Tokens.Struct); }
                else if (temp1[i] == "arr") { tokens.Add(Tokens.Array); }

                else if (temp1[i] == "while") { tokens.Add(Tokens.Class); }
                else if (temp1[i] == "for") { tokens.Add(Tokens.Class); }

                else if (temp1[i] == "if") { tokens.Add(Tokens.Class); }
                else if (temp1[i] == "elif") { tokens.Add(Tokens.Class); }
                else if (temp1[i] == "else") { tokens.Add(Tokens.Class); }

                else if (temp1[i] == "new") { tokens.Add(Tokens.New); }
                else if (temp1[i] == "=") { tokens.Add(Tokens.Assign); }
                else if (temp1[i] == "void") { tokens.Add(Tokens.Void); }
                else if (temp1[i] == "int") { tokens.Add(Tokens.Integer); }
                else if (temp1[i] == "float") { tokens.Add(Tokens.Float); }
                else if (temp1[i] == "char") { tokens.Add(Tokens.Char); }
                else if (temp1[i] == ",") { tokens.Add(Tokens.Comma); }

                else if (temp1[i] == "+") { tokens.Add(Tokens.Add); }
                else if (temp1[i] == "-") { tokens.Add(Tokens.Subtract); }
                else if (temp1[i] == "*") { tokens.Add(Tokens.Multiple); }
                else if (temp1[i] == "/") { tokens.Add(Tokens.Divide); }

                else if (temp1[i] == "==") { tokens.Add(Tokens.Equal); }
                else if (temp1[i] == "!=") { tokens.Add(Tokens.IsNot); }
                else if (temp1[i] == ">=") { tokens.Add(Tokens.MoreThan); }
                else if (temp1[i] == "<=") { tokens.Add(Tokens.LessThan); }

                else if (temp1[i] == ";") { tokens.Add(Tokens.Semicolon); }
                else if (temp1[i] == "[") { tokens.Add(Tokens.OpenBracket); }
                else if (temp1[i] == "]") { tokens.Add(Tokens.CloseBracket); }
                else if (temp1[i] == "(") { tokens.Add(Tokens.OpenParentheses); }
                else if (temp1[i] == ")") { tokens.Add(Tokens.CloseParentheses); }
                else if (temp1[i] == "{") { tokens.Add(Tokens.OpenCurlyBracket); }

                else if (temp1[i] == "}") { tokens.Add(Tokens.CloseCurlyBracket); }
                else { tokens.Add(Tokens.Value); values.Add(temp1[i]); }
            }

            return tokens;
        }
    }
}

File: Silent/SilentCompiler/Lexer.c
#include "Lexer.h"

void clearCodeBuffer(char* buffer, int *bufferPointer) 
{

	for (int i = 0; i <= *bufferPointer; i++) 
	{
		*(buffer + i) = 0;
	}

	bufferPointer = 0;

}

char addToken(Program *program, char *token, int *bufferPointer)
{
	if (bufferPointer == 0) { return -1; }
	token[*bufferPointer] = '\0';

	printf(token);

	if (!strcmp(token, "public")) 
	{
		*(program->tokens + program->tokensPointer) =  Public; 
		program->tokensPointer++;
	}

	else if (!strcmp(token, "private")) 
	{
		*(program->tokens + program->tokensPointer) = Private; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "protected")) 
	{
		*(program->tokens + program->tokensPointer) = Protected; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "static")) 
	{
		*(program->tokens + program->tokensPointer) = Static; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "namespace")) 
	{ 
		*(program->tokens + program->tokensPointer) = Namespace; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "using")) 
	{
		*(program->tokens + program->tokensPointer) = Using; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "class")) 
	{
		*(program->tokens + program->tokensPointer) = Class; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "method")) 
	{
		*(program->tokens + program->tokensPointer) = Method;
		program->tokensPointer++;
	}
	else if (!strcmp(token, "function")) 
	{
		*(program->tokens + program->tokensPointer) = Function; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "return")) 
	{
		*(program->tokens + program->tokensPointer) = Return; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "struct")) 
	{
		*(program->tokens + program->tokensPointer) = Struct; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "array")) 
	{
		*(program->tokens + program->tokensPointer) = Array; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "while")) 
	{
		*(program->tokens + program->tokensPointer) = While; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "for")) 
	{
		*(program->tokens + program->tokensPointer) = For; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "if")) 
	{
		*(program->tokens + program->tokensPointer) = If; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "elif")) 
	{
		*(program->tokens + program->tokensPointer) = Elif; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "else")) 
	{
		*(program->tokens + program->tokensPointer) = Else; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "new")) 
	{ 
		*(program->tokens + program->tokensPointer) = New; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "void")) 
	{
		*(program->tokens + program->tokensPointer) = Void; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "int")) 
	{
		*(program->tokens + program->tokensPointer) = Integer; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "float")) 
	{
		*(program->tokens + program->tokensPointer) = Float; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "string")) 
	{
		*(program->tokens + program->tokensPointer) = String; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "char")) 
	{
		*(program->tokens + program->tokensPointer) = Char; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "var")) 
	{
		*(program->tokens + program->tokensPointer) = Variable; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "==")) 
	{
		*(program->tokens + program->tokensPointer) = Equal; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "!=")) 
	{
		*(program->tokens + program->tokensPointer) = IsNot; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, ">=")) 
	{ 
		*(program->tokens + program->tokensPointer) = MoreThan; 
		program->tokensPointer++;
	}
	else if (!strcmp(token, "<=")) 
	{
		*(program->tokens + program->tokensPointer) = LessThan; 
		program->tokensPointer++;
	}
	else 
	{
		*(program->tokens + program->tokensPointer) = Value;
		program->tokensPointer++;
		program->values[program->valuePointer] =
			*(program->tokens + program->tokensPointer);
	}
	clearCodeBuffer(token, bufferPointer);

}

void AppendToBuffer(char* buffer, unsigned int *bufferPointer, char appendData) 
{
	buffer[*bufferPointer] = appendData;
	*bufferPointer++;
}

void Tokenize(Program *program)
{
	//allocate scanning variables
	char* CodeBuffer = malloc(50);
	unsigned int CodeBufferPointer = 0;
	char scanning = 1;

	//main scanning loop
	for (unsigned long i = 0; i <= program->sourceLength; i++)
	{

		if (!scanning)
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			scanning = 1;
		}

		//check for spaces
		else if (*(program->source + i) == ' ') { scanning = 0; continue; }
		else if (*(program->source + i) == '\t') { scanning = 0; continue; }
		else if (*(program->source + i) == '\n') { scanning = 0; continue; }
		else if (*(program->source + i) == '\r') { scanning = 0; continue; }
		else if (*(program->source + i) == '\f') { scanning = 0; continue; }
		else if (*(program->source + i) == '\v') { scanning = 0; continue; }

		else if (*(program->source + i) == '.')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = Fullstop;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '=')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = Assign;
			program->tokensPointer++;
		}
		else if (*(program->source + i) == ',')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = Comma;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '+')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = Add;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '-')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = Subtract;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '*')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = Asterisk;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '//')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = Divide;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == ';')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = Semicolon;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '[')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = OpenBracket;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == ']')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = CloseBracket;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '(')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = OpenParentheses;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == ')')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = CloseParentheses;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '{')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = OpenCurlyBracket;
			program->tokensPointer++;
		}

		else if (*(program->source + i) == '}')
		{
			addToken(program, CodeBuffer, &CodeBufferPointer);
			*(program->tokens + program->tokensPointer) = CloseCurlyBracket;
			program->tokensPointer++;
		}

		//if the value is not a token add it to the buffer
		else
		{
			*(CodeBuffer + CodeBufferPointer) = *(program->source + i);
			CodeBufferPointer++;
		}
	}
}

